From: Folia Compatibility Patch
Subject: [PATCH 0024] FireballEffect: Migrate to teleportAsync

AE PATCH REASON: Uses synchronous teleport on fireball entity
AE PATCH FIX: Migrate to teleportAsync
AE PATCH RISK: MEDIUM
AE PATCH EXCLUDES: net/advancedplugins/ae/impl/effects/effects/effects/internal/FireballEffect.class, net/advancedplugins/ae/impl/effects/effects/effects/internal/FireballEffect$*.class

---
diff --git a/net/advancedplugins/ae/impl/effects/effects/effects/internal/FireballEffect.java b/net/advancedplugins/ae/impl/effects/effects/effects/internal/FireballEffect.java
index a2e7414..cf5e61a 100644
--- a/net/advancedplugins/ae/impl/effects/effects/effects/internal/FireballEffect.java
+++ b/net/advancedplugins/ae/impl/effects/effects/effects/internal/FireballEffect.java
@@ -11,32 +11,32 @@ import org.bukkit.plugin.java.JavaPlugin;
 import org.bukkit.util.Vector;
 
 public class FireballEffect extends AdvancedEffect {
-   public FireballEffect(JavaPlugin var1) {
-      super(var1, "FIREBALL", "Shoot a fireball", "%e");
+   public FireballEffect(JavaPlugin plugin) {
+      super(plugin, "FIREBALL", "Shoot a fireball", "%e");
    }
 
-   public boolean executeEffect(ExecutionTask var1, LivingEntity var2, String[] var3) {
-      return this.executeEffect(var1, var2.getLocation(), var3);
+   public boolean executeEffect(ExecutionTask task, LivingEntity entity, String[] args) {
+      return this.executeEffect(task, entity.getLocation(), args);
    }
 
-   public boolean executeEffect(ExecutionTask var1, Location var2, String[] var3) {
-      Fireball var4 = (Fireball)var2.getWorld().spawnEntity(var2, EntityType.FIREBALL);
-      var4.setIsIncendiary(false);
-      var4.setYield(0.0F);
-      Vector var5 = new Vector(var2.getX(), var2.getY(), var2.getZ());
-      Vector var6 = new Vector(var4.getLocation().getX(), var4.getLocation().getY(), var4.getLocation().getZ());
-      if (ASManager.isExcessVelocity(var5.subtract(var6))) {
+   public boolean executeEffect(ExecutionTask task, Location location, String[] args) {
+      Fireball fireball = (Fireball)location.getWorld().spawnEntity(location, EntityType.FIREBALL);
+      fireball.setIsIncendiary(false);
+      fireball.setYield(0.0F);
+      Vector target = new Vector(location.getX(), location.getY(), location.getZ());
+      Vector current = new Vector(fireball.getLocation().getX(), fireball.getLocation().getY(), fireball.getLocation().getZ());
+      if (ASManager.isExcessVelocity(target.subtract(current))) {
          return false;
       } else {
-         var4.setVelocity(var5.subtract(var6));
-         var4.teleport(var4.getLocation());
+         fireball.setVelocity(target.subtract(current));
+         fireball.teleportAsync(fireball.getLocation());
 
          try {
-            var4.setVelocity(var5.subtract(var6));
-         } catch (Exception var8) {
+            fireball.setVelocity(target.subtract(current));
+         } catch (Exception e) {
          }
 
-         var4.setYield(0.0F);
+         fireball.setYield(0.0F);
          return true;
       }
    }

